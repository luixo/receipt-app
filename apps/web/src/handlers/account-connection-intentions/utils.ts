import { TRPCError } from "@trpc/server";
import type { z } from "zod";

import { ACCOUNT_CONNECTIONS_INTENTIONS } from "~db/consts";
import type { AccountsId, UsersId } from "~db/models";
import type { Database } from "~db/types";
import type { emailSchema } from "~web/handlers/validation";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: z.infer<typeof emailSchema>,
	asUserId: UsersId,
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select(["id", "avatarUrl"])
		.where("email", "=", toEmail.lowercase)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `Account with email "${toEmail.original}" does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail.lowercase)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId),
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `Account with email "${toEmail.original}" is already connected to user "${connectedUser.name}".`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where((eb) =>
			eb.and({
				accountId: targetAccount.id,
				targetAccountId: fromAccountId,
			}),
		)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where((eb) =>
					eb.and({
						ownerAccountId: targetAccount.id,
						id: viceVersaIntention.userId,
					}),
				)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where((eb) =>
					eb.and({
						ownerAccountId: fromAccountId,
						id: asUserId,
					}),
				)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: targetAccount.id,
						targetAccountId: fromAccountId,
					}),
				)
				.executeTakeFirst();
		});
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				createdAt: new Date(),
			})
			.executeTakeFirst();
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: fromAccountId,
						targetAccountId: targetAccount.id,
					}),
				)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await database
				.selectFrom("users")
				.select("name")
				.where("id", "=", existingIntention.userId)
				.executeTakeFirstOrThrow();
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${toEmail.original}" as user "${existingUser.name}".`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${user.name}".`,
			});
			// This is probably a c8 bug
			/* c8 ignore next */
		}
		/* c8 ignore next 2 */
		throw e;
	}
};
