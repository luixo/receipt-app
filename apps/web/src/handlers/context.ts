import { type H3Event, createEvent } from "@tanstack/react-start/server";
import type { inferProcedureBuilderResolverOptions } from "@trpc/server";
import type { TRPCRequestInfo } from "@trpc/server/http";
import type { IncomingMessage, ServerResponse } from "node:http";

import type { Database } from "~db/database";
import type { TestContext } from "~tests/backend/utils/test";
import type { authProcedure } from "~web/handlers/trpc";
import type { CacheDbOptions } from "~web/providers/cache-db";
import type { EmailOptions } from "~web/providers/email";
import type { ExchangeRateOptions } from "~web/providers/exchange-rate";
import type { Logger } from "~web/providers/logger";
import type { S3Options } from "~web/providers/s3";

type TestContextPicks = Pick<TestContext, "getSalt" | "getUuid"> & {
	database: Database;
	logger: Logger;
	emailOptions: EmailOptions;
	cacheDbOptions: CacheDbOptions;
	exchangeRateOptions: ExchangeRateOptions;
	s3Options: S3Options;
};

export type NetContext = {
	event: H3Event;
	info: TRPCRequestInfo;
};

export type UnauthorizedContext = NetContext & TestContextPicks;

export type AuthorizedContext = inferProcedureBuilderResolverOptions<
	typeof authProcedure
>["ctx"];

export const createContext = (
	{
		req,
		res,
		info,
	}: { req: IncomingMessage; res: ServerResponse; info: TRPCRequestInfo },
	testContext: TestContextPicks,
): UnauthorizedContext => ({
	info,
	event: createEvent(req, res),
	...testContext,
});
