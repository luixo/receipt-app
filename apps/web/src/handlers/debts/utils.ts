import type { Database, DebtsId, SimpleInsertObject } from "~db";
import { nonNullishGuard } from "~utils";

export const upsertAutoAcceptedDebts = async (
	database: Database,
	debts: (SimpleInsertObject<"debts"> & { isNew: boolean })[],
) => {
	const fetchedDebts = await database
		.selectFrom("debts")
		.where((eb) =>
			eb.or(
				debts.map((debt) =>
					!debt.isNew
						? eb.and({
								id: debt.id,
								ownerAccountId: debt.ownerAccountId,
						  })
						: debt.receiptId
						? eb.and({
								ownerAccountId: debt.ownerAccountId,
								userId: debt.userId,
								receiptId: debt.receiptId,
						  })
						: eb("debts.ownerAccountId", "is", null),
				),
			),
		)

		.select([
			"debts.id",
			"debts.ownerAccountId",
			"debts.userId",
			"debts.receiptId",
		])
		.execute();
	const counterpartyDebts = debts.map((debt) =>
		fetchedDebts.find((fetchedDebt) => {
			if (debt.isNew) {
				if (!debt.receiptId || !fetchedDebt.receiptId) {
					return false;
				}
				return (
					fetchedDebt.ownerAccountId === debt.ownerAccountId &&
					fetchedDebt.userId === debt.userId &&
					fetchedDebt.receiptId === debt.receiptId
				);
			}
			return (
				fetchedDebt.ownerAccountId === debt.ownerAccountId &&
				fetchedDebt.id === debt.id
			);
		}),
	);
	const existentDebts = debts
		.map((nextDebt, index) =>
			counterpartyDebts[index]
				? ([nextDebt, counterpartyDebts[index]!] as const)
				: null,
		)
		.filter(nonNullishGuard);
	const nonExistentDebts = debts
		.map((nextDebt, index) => (counterpartyDebts[index] ? null : nextDebt))
		.filter(nonNullishGuard);
	const [newDebts, ...updatedDebts] = await Promise.all([
		nonExistentDebts.length === 0
			? ([] as { id: DebtsId }[])
			: database
					.insertInto("debts")
					.values(nonExistentDebts.map(({ isNew, ...debt }) => debt))
					.returning(["debts.id"])
					.execute(),
		...existentDebts.map(([nextDebt, currentDebt]) =>
			database
				.updateTable("debts")
				.set({
					amount: nextDebt.amount,
					timestamp: nextDebt.timestamp,
					currencyCode: nextDebt.currencyCode,
					receiptId: nextDebt.receiptId,
					lockedTimestamp: nextDebt.lockedTimestamp,
				})
				.where((eb) =>
					eb.and({
						id: currentDebt.id,
						ownerAccountId: currentDebt.ownerAccountId,
						userId: currentDebt.userId,
					}),
				)
				.returning(["debts.id", "debts.receiptId"])
				.executeTakeFirst(),
		),
	]);
	return { newDebts, updatedDebts };
};
