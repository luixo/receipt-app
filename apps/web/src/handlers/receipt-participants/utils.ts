import type { Selection } from "kysely";

import type { ReceiptsId, UsersId } from "~db/models";
import type {
	Database,
	ReceiptsDatabase,
	ReceiptsSelectExpression,
} from "~db/types";

export const getReceiptParticipant = <
	SE extends ReceiptsSelectExpression<"receiptParticipants">,
>(
	database: Database,
	userId: UsersId,
	receiptId: ReceiptsId,
	selectExpression: SE[],
): Promise<
	Selection<ReceiptsDatabase, "receiptParticipants", SE> | undefined
> =>
	database
		.selectFrom("receiptParticipants")
		.where((eb) => eb.and({ receiptId, userId }))
		.select(selectExpression)
		.limit(1)
		.executeTakeFirst();
