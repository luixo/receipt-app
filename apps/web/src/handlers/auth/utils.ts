import { TRPCError } from "@trpc/server";

import { DAY } from "~utils";
import type { AccountsId, SessionsSessionId } from "~web/db/models";
import { generateConfirmEmailEmail } from "~web/email/utils";
import type { UnauthorizedContext } from "~web/handlers/context";
import { getEmailClient } from "~web/providers/email";

// How long a session should last
export const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long session should be intact before we update it
export const SESSION_SHOULD_UPDATE_EVERY = 2 * DAY;

export const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);

export const createAuthorizationSession = async (
	ctx: UnauthorizedContext,
	accountId: AccountsId,
) => {
	const uuid: SessionsSessionId = ctx.getUuid();
	const expirationDate = getExpirationDate();
	await ctx.database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const sendVerificationEmail = async (
	ctx: UnauthorizedContext,
	email: string,
	token: string,
) => {
	if (!ctx.emailOptions.active) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient(ctx).send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(ctx, token),
		});
	} catch (e) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				/* c8 ignore next */
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
