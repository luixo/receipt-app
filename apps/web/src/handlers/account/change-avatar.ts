import { TRPCError } from "@trpc/server";
import Sharp from "sharp";

import { avatarFormSchema } from "~app/utils/validation";
import { MAX_AVATAR_BYTESIZE, MAX_AVATAR_SIDE_SIZE } from "~utils/images";
import { authProcedure } from "~web/handlers/trpc";
import { getS3Client } from "~web/providers/s3";

export const S3_AVATAR_PREFIX = "avatars";

const ALLOWED_FORMATS: (keyof Sharp.FormatEnum)[] = ["png", "jpeg", "jpg"];

const validateImage = async (image: Buffer) => {
	const parsedImage = Sharp(image);
	const metadata = await parsedImage.metadata();
	if (!metadata.size || metadata.size > MAX_AVATAR_BYTESIZE) {
		throw new TRPCError({
			code: "BAD_REQUEST",
			message: `Maximum bytesize allowed is ${MAX_AVATAR_BYTESIZE}.`,
		});
	}
	if (!metadata.format || !ALLOWED_FORMATS.includes(metadata.format)) {
		throw new TRPCError({
			code: "BAD_REQUEST",
			/* c8 ignore next */
			message: `Format "${metadata.format || "unknown"}" is not allowed.`,
		});
	}
	const maxSizeAllowed = MAX_AVATAR_SIDE_SIZE;
	if (!metadata.height || metadata.height > maxSizeAllowed) {
		throw new TRPCError({
			code: "BAD_REQUEST",
			message: `Maximum height allowed is ${maxSizeAllowed}.`,
		});
	}
	if (!metadata.width || metadata.width > maxSizeAllowed) {
		throw new TRPCError({
			code: "BAD_REQUEST",
			message: `Maximum width allowed is ${maxSizeAllowed}.`,
		});
	}
	if (metadata.width !== metadata.height) {
		throw new TRPCError({
			code: "BAD_REQUEST",
			message: `Expected to have equal height and width, got ${metadata.width}x${metadata.height}.`,
		});
	}
	return parsedImage.toBuffer();
};

export const procedure = authProcedure
	.input(avatarFormSchema)
	.mutation(async ({ input, ctx }) => {
		const { database } = ctx;
		if (!input.avatar) {
			await database
				.updateTable("accounts")
				.set({ avatarUrl: null })
				.where("accounts.id", "=", ctx.auth.accountId)
				.executeTakeFirst();
			return;
		}
		const s3Client = getS3Client(ctx);
		const validatedImage = await validateImage(
			Buffer.from(await input.avatar.arrayBuffer()),
		);
		const avatarKey = [S3_AVATAR_PREFIX, `${ctx.auth.accountId}.png`].join("/");
		await s3Client.putObject(avatarKey, validatedImage);
		const url = `${[s3Client.endpoint, s3Client.bucket, avatarKey].join(
			"/",
		)}?lastModified=${Date.now()}`;
		await database
			.updateTable("accounts")
			.set({ avatarUrl: url })
			.where("accounts.id", "=", ctx.auth.accountId)
			.executeTakeFirst();
		return { url };
	});
