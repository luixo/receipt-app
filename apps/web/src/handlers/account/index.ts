import { procedure as changeAvatar } from "./change-avatar";
import { procedure as changeName } from "./change-name";
import { procedure as changePassword } from "./change-password";
import { procedure as get } from "./get";
import { procedure as logout } from "./logout";
import { procedure as resendEmail } from "./resend-email";

export const router = {
	get,
	logout,
	changeName,
	changePassword,
	changeAvatar,
	resendEmail,
};
