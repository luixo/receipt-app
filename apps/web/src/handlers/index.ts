import { t } from "~web/handlers/trpc";

import { router as accountRouter } from "./account/index";
import { router as accountConnectionIntentionsRouter } from "./account-connection-intentions/index";
import { router as accountSettingsRouter } from "./account-settings/index";
import { router as adminRouter } from "./admin/index";
import { router as authRouter } from "./auth/index";
import { router as currencyRouter } from "./currency/index";
import { router as debtIntentionsRouter } from "./debt-intentions/index";
import { router as debtsRouter } from "./debts/index";
import { router as receiptItemConsumersRouter } from "./receipt-item-consumers/index";
import { router as receiptItemPayersRouter } from "./receipt-item-payers/index";
import { router as receiptItemsRouter } from "./receipt-items/index";
import { router as receiptParticipantsRouter } from "./receipt-participants/index";
import { router as receiptsRouter } from "./receipts/index";
import { router as resetPasswordIntentionsRouter } from "./reset-password-intentions/index";
import { router as sessionsRouter } from "./sessions/index";
import { router as usersRouter } from "./users/index";
import { router as utilsRouter } from "./utils/index";

export const router = t.router({
	// No auth
	sessions: sessionsRouter,
	auth: authRouter,
	resetPasswordIntentions: resetPasswordIntentionsRouter,
	utils: utilsRouter,
	// Auth
	account: accountRouter,
	accountSettings: accountSettingsRouter,
	receipts: receiptsRouter,
	receiptItems: receiptItemsRouter,
	users: usersRouter,
	receiptParticipants: receiptParticipantsRouter,
	currency: currencyRouter,
	receiptItemConsumers: receiptItemConsumersRouter,
	receiptItemPayers: receiptItemPayersRouter,
	accountConnectionIntentions: accountConnectionIntentionsRouter,
	debts: debtsRouter,
	debtIntentions: debtIntentionsRouter,
	// Admin
	admin: adminRouter,
});
