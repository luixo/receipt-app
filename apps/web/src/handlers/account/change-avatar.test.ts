import Sharp from "sharp";
import { assert, describe, expect } from "vitest";

import { createAuthContext } from "~tests/backend/utils/context";
import { insertAccountWithSession } from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "~tests/backend/utils/expect";
import { test } from "~tests/backend/utils/test";
import { MAX_AVATAR_BYTESIZE, MAX_AVATAR_SIDE_SIZE } from "~utils/images";
import { t } from "~web/handlers/trpc";

import { S3_AVATAR_PREFIX, procedure } from "./change-avatar";

const createCaller = t.createCallerFactory(t.router({ procedure }));

const getFormData = (bits?: BlobPart[]) => {
	const formData = new FormData();
	if (bits) {
		formData.append("avatar", new File(bits, "avatar.png"));
	}
	return formData;
};

type FormImageOptions = {
	exif?: Record<string, unknown>;
	format?: "png" | "jpeg" | "webp";
	type?: "static" | "noise";
};
const defaultSettings: FormImageOptions = {
	format: "png",
};

const generateFormWithImage = async (
	width: number,
	height: number,
	{ exif, format = "png", type = "static" }: FormImageOptions = defaultSettings,
) => {
	let image = Sharp({
		create: {
			width,
			height,
			channels: 4,
			background: { r: 255, g: 0, b: 0, alpha: 0.5 },
			noise:
				type === "noise"
					? {
							type: "gaussian" as const,
							mean: 128,
							sigma: 30,
						}
					: undefined,
		},
	});
	image = image[format]();
	if (exif) {
		image = image.withMetadata({ exif });
	}
	const buffer = await image.toBuffer();
	return getFormData([buffer]);
};

describe("account.changeAvatar", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			createCaller(context).procedure(getFormData([])),
		);

		describe("avatar", () => {
			test("is too big in bytes", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					async () =>
						caller.procedure(
							await generateFormWithImage(
								MAX_AVATAR_SIDE_SIZE * 3,
								MAX_AVATAR_SIDE_SIZE * 3,
								{ type: "noise" },
							),
						),
					"BAD_REQUEST",
					`Maximum bytesize allowed is ${MAX_AVATAR_BYTESIZE}.`,
				);
			});

			test("is too tall", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					async () =>
						caller.procedure(
							await generateFormWithImage(
								MAX_AVATAR_SIDE_SIZE,
								MAX_AVATAR_SIDE_SIZE + 1,
							),
						),
					"BAD_REQUEST",
					`Maximum height allowed is ${MAX_AVATAR_SIDE_SIZE}.`,
				);
			});

			test("is too wide", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					async () =>
						caller.procedure(
							await generateFormWithImage(
								MAX_AVATAR_SIDE_SIZE + 1,
								MAX_AVATAR_SIDE_SIZE,
							),
						),
					"BAD_REQUEST",
					`Maximum width allowed is ${MAX_AVATAR_SIDE_SIZE}.`,
				);
			});

			test("is not square", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					async () =>
						caller.procedure(
							await generateFormWithImage(
								MAX_AVATAR_SIDE_SIZE,
								MAX_AVATAR_SIDE_SIZE - 1,
							),
						),
					"BAD_REQUEST",
					`Expected to have equal height and width, got ${MAX_AVATAR_SIDE_SIZE}x${
						MAX_AVATAR_SIDE_SIZE - 1
					}.`,
				);
			});

			test("is not of allowed format", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					async () =>
						caller.procedure(
							await generateFormWithImage(
								MAX_AVATAR_SIDE_SIZE,
								MAX_AVATAR_SIDE_SIZE,
								{ format: "webp" },
							),
						),
					"BAD_REQUEST",
					`Format "webp" is not allowed.`,
				);
			});
		});

		test("provider broken", async ({ ctx }) => {
			ctx.s3Options.broken = true;
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));
			await expectTRPCError(
				async () =>
					caller.procedure(
						await generateFormWithImage(
							MAX_AVATAR_SIDE_SIZE,
							MAX_AVATAR_SIDE_SIZE,
						),
					),
				"INTERNAL_SERVER_ERROR",
				"Test context broke s3 service error",
			);
		});
	});

	describe("functionality", () => {
		test("avatar changes to null", async ({ ctx }) => {
			// Verifying other users are not affected
			await insertAccountWithSession(ctx);
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));

			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure(getFormData()),
			);
			expect(result).toBeUndefined();
			expect(ctx.s3Options.mock.getMessages()).toHaveLength(0);
		});

		test("avatar changes to a given image", async ({ ctx }) => {
			// Verifying other users are not affected
			await insertAccountWithSession(ctx);
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));

			const result = await expectDatabaseDiffSnapshot(ctx, async () =>
				caller.procedure(
					await generateFormWithImage(
						MAX_AVATAR_SIDE_SIZE,
						MAX_AVATAR_SIDE_SIZE,
					),
				),
			);
			const key = [S3_AVATAR_PREFIX, `${accountId}.png`].join("/");
			const url = `${[
				ctx.s3Options.mock.endpoint,
				ctx.s3Options.mock.bucket,
				key,
			].join(
				"/",
				// eslint-disable-next-line no-restricted-syntax
			)}?lastModified=${Date.now()}`;
			expect(result).toEqual({ url });
			expect(ctx.s3Options.mock.getMessages()).toHaveLength(1);
			const message = ctx.s3Options.mock.getMessages()[0];
			assert(message);
			expect(message.objectLength).toBeGreaterThan(1000);
			expect(message.objectLength).toBeLessThan(5000);
			expect(message).toStrictEqual<typeof message>({
				key,
				objectLength: message.objectLength,
			});
		});

		test("metadata is stripped", async ({ ctx }) => {
			// Verifying other users are not affected
			await insertAccountWithSession(ctx);
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));

			await caller.procedure(
				await generateFormWithImage(MAX_AVATAR_SIDE_SIZE, MAX_AVATAR_SIDE_SIZE),
			);
			await caller.procedure(
				await generateFormWithImage(
					MAX_AVATAR_SIDE_SIZE,
					MAX_AVATAR_SIDE_SIZE,
					{ exif: { IFD0: { Copyright: "Anything" } } },
				),
			);
			const getObjectLength = (index: number) => {
				const message = ctx.s3Options.mock.getMessages()[index];
				assert(message);
				return message.objectLength;
			};
			expect(getObjectLength(0)).toEqual(getObjectLength(1));
		});
	});
});
