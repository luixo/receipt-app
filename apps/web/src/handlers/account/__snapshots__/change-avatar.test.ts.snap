// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`account.changeAvatar > functionality > avatar changes to a given image 1`] = `
{
  "updated": {
    "accounts": {
      "id:6fbc201e-126e-4834-8710-3c2910e12e39": {
        "avatarUrl": "https://fake-endpoint.org/test-bucket/avatars/6fbc201e-126e-4834-8710-3c2910e12e39.png?lastModified=1577836800000",
        "updatedAt": "2020-01-01 00:01:00+00",
      },
    },
  },
}
`;

exports[`account.changeAvatar > functionality > avatar changes to null 1`] = `
{
  "updated": {
    "accounts": {
      "id:fd2e5e36-7dd4-49d4-a77f-058e4840222b": {
        "avatarUrl": null,
        "updatedAt": "2020-01-01 00:01:00+00",
      },
    },
  },
}
`;
